core/src/main/java/org/apache/calcite/rel/metadata/RelMdForeignKeys.java

```diff
@@ -0,0 +1,246 @@
+      return EMPTY_BIT_SET;
+    }
+    final ImmutableBitSet inputForeignKeys = mq.getForeignKeys(rel.getInput(), ignoreNulls);
+    if (inputForeignKeys.isEmpty()) {
+      return EMPTY_BIT_SET;
+    }
+    return groupSet.intersect(inputForeignKeys);
+  }
+
+  public ImmutableBitSet getForeignKeys(Project rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+    return getProjectForeignKeys(rel, mq, ignoreNulls, rel.getProjects());
+  }
+
+  public ImmutableBitSet getForeignKeys(Calc rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+    RexProgram program = rel.getProgram();
+    return getProjectForeignKeys(rel, mq, ignoreNulls,
+        Util.transform(program.getProjectList(), program::expandLocalRef));
+  }
+
+  private static ImmutableBitSet getProjectForeignKeys(SingleRel rel, RelMetadataQuery mq,
+      boolean ignoreNulls,
+      List<RexNode> projExprs) {
+
+    // Single input can be mapped to multiple outputs
+    final ImmutableListMultimap.Builder<Integer, Integer> inToOutIndexBuilder =
+        ImmutableListMultimap.builder();
+    final ImmutableBitSet.Builder inColumnsBuilder = ImmutableBitSet.builder();
+    for (int i = 0; i < projExprs.size(); i++) {
+      RexNode projExpr = projExprs.get(i);
+      if (projExpr instanceof RexInputRef) {
+        int inputIndex = ((RexInputRef) projExpr).getIndex();
+        inToOutIndexBuilder.put(inputIndex, i);
+        inColumnsBuilder.set(inputIndex);
+      }
+    }
+    final ImmutableBitSet inColumnsUsed = inColumnsBuilder.build();
+    if (inColumnsUsed.isEmpty()) {
+      return EMPTY_BIT_SET;
+    }
+
+    final Map<Integer, ImmutableBitSet> mapInToOutPos =
+        Maps.transformValues(inToOutIndexBuilder.build().asMap(), ImmutableBitSet::of);
+    final ImmutableBitSet inputForeignKeys =
+        mq.getForeignKeys(rel.getInput(), ignoreNulls);
+    if (inputForeignKeys.isEmpty()) {
+      return EMPTY_BIT_SET;
+    }
+
+    ImmutableBitSet allOutColumns = ImmutableBitSet.of();
+    for (int index : inputForeignKeys.asList()) {
+      ImmutableBitSet outColumns = mapInToOutPos.get(index);
+      if (outColumns != null && !outColumns.isEmpty()) {
+        allOutColumns = allOutColumns.union(outColumns);
+      }
+    }
+    return allOutColumns.isEmpty() ? EMPTY_BIT_SET : allOutColumns;
+  }
+
+  public ImmutableBitSet getForeignKeys(TableScan rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+    final RelOptTable table = rel.getTable();
+    final BuiltInMetadata.ForeignKeys.Handler handler =
+        table.unwrap(BuiltInMetadata.ForeignKeys.Handler.class);
+    if (handler != null) {
+      return handler.getForeignKeys(rel, mq, ignoreNulls);
+    }
+
+    final List<RelReferentialConstraint> referentialConstraints =
+        table.getReferentialConstraints();
+    if (referentialConstraints == null || referentialConstraints.size() == 0) {
+      return EMPTY_BIT_SET;
+    }
+    final List<IntPair> foreignUniquePair = referentialConstraints.stream()
```

Comment:
If you really want unique, you have to remove duplicates (using a `Set<IntPair>` possibly).

If you don't need unique, please update the field name not to confuse readers.

Reply:
OK, got your point. 
I think we can represent multi foreign keys in a  bit set if a table has multi referential constraints. 
And retain the right index position. 
For example:
`SELECT ename, deptno, deptno
FROM emp `
where deptno is foreign key. 
We should returen {1, 2} when call getForeignKeys method.